src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,401 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Pushes projections containing unique columns down through other operators.
+//!
+//! This action improves the quality of the query, in that projections without
+//! repeated columns reduce the width of data in the dataflow.
+//!
+//! Some comments have been inherited from the `Demand` transform.
+
+use std::collections::{HashMap, HashSet};
+
+use expr::{Id, JoinInputMapper, MirRelationExpr, MirScalarExpr};
+
+use crate::TransformArgs;
+
+/// Pushes projections down through other operators.
+#[derive(Debug)]
+pub struct ProjectionPushdown;
+
+impl crate::Transform for ProjectionPushdown {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        self.action(
+            relation,
+            (0..relation.arity()).collect(),
+            &mut HashMap::new(),
+        );
+        Ok(())
+    }
+}
+
+impl ProjectionPushdown {
+    /// Pushes the `desired_projection` down through `relation`.
+    ///
+    /// This action transforms `relation` to a `MirRelationExpr` equivalent to
+    /// `relation.project(desired_projection)`.
+    ///
+    /// `desired_projection` is expected to consist of unique columns.
+    fn action(
+        &self,
+        relation: &mut MirRelationExpr,
+        mut desired_projection: Vec<usize>,
+        gets: &mut HashMap<Id, HashSet<usize>>,
+    ) {
+        // Try to push the desired projection down through `relation`.
+        // In the process `relation` is transformed to a ` MirRelationExpr`
+        // equivalent to `relation.project(actual_projection)`.
+        let actual_projection = match relation {
+            // Nothing can be pushed through leaf nodes.
+            MirRelationExpr::Constant { .. } => (0..relation.arity()).collect(),
+            MirRelationExpr::Get { id, .. } => {
+                gets.entry(*id)
+                    .or_insert_with(HashSet::new)
+                    .extend(desired_projection.clone());
```

Comment:
I think this could be `desired_projection.iter().cloned()` to avoid an allocation, fwiw.